src/subdataframe/subdataframe.jl

```diff
@@ -176,32 +176,106 @@ Base.@propagate_inbounds Base.getindex(df::SubDataFrame, row_ind::typeof(!),
+        if !(val isa AbstractVector && nrow(sdf) == length(val))
+            throw(ArgumentError("Assigned value must be a vector with length " *
+                                "equal to number of rows in the SubDataFrame"))
+        end
+        T = eltype(val)
+        newcol = Tables.allocatecolumn(Union{T, Missing}, nrow(parent(sdf)))
+        fill!(newcol, missing)
+        newcol[rows(sdf)] = val
+        parent(sdf)[!, colinds] = newcol
+    else
+        parent(sdf)[rows(sdf), parentcols(index(sdf), colinds)] = val
+    end
+    return sdf
+end
+
+function Base.setindex!(sdf::SubDataFrame, v::AbstractVector,
+                        ::typeof(!), col_ind::ColumnIndex)
+    if col_ind isa Union{Signed, Unsigned} && !(1 <= col_ind <= ncol(sdf))
+        throw(ArgumentError("Cannot assign to non-existent column: $col_ind"))
+    end
+    if col_ind isa SymbolOrString && columnindex(sdf, col_ind) == 0
+        if !is_column_adding_allowed(sdf)
+            throw(ArgumentError("creating new columns in a SubDataFrame that subsets " *
+                                "columns of its parent data frame is disallowed"))
+        end
+        sdf[:, col_ind] = v
+    else
+        pdf = parent(sdf)
+        p_col_ind = parentcols(index(sdf), col_ind)
+        old_col = pdf[!, p_col_ind]
+        T = eltype(old_col)
+        S = eltype(v)
+        newcol = Tables.allocatecolumn(promote_type(T, S), length(old_col))
+        newcol .= old_col
+        newcol[rows(sdf)] = v
+        pdf[!, p_col_ind] = newcol
+    end
     return sdf
 end
-Base.@propagate_inbounds function Base.setindex!(sdf::SubDataFrame, val::Any, ::typeof(!), colinds::Any)
-    throw(ArgumentError("setting index of SubDataFrame using ! as row selector is not allowed"))
+
+for T in MULTICOLUMNINDEX_TUPLE
+    @eval function Base.setindex!(sdf::SubDataFrame,
+                                  new_df::AbstractDataFrame,
+                                  row_inds::typeof(!),
+                                  col_inds::$T)
+        idxs = index(sdf)[col_inds]
+        if view(_names(sdf), idxs) != _names(new_df)
+            throw(ArgumentError("Column names in source and target data frames do not match"))
+        end
+        for (j, col) in enumerate(idxs)
+            sdf[!, col] = new_df[!, j]
+        end
+        return sdf
+    end
+
+    @eval function Base.setindex!(sdf::SubDataFrame,
+                                  mx::AbstractMatrix,
+                                  row_inds::typeof(!),
+                                  col_inds::$T)
+        idxs = index(sdf)[col_inds]
+        if size(mx, 2) != length(idxs)
+            throw(DimensionMismatch("number of selected columns ($(length(idxs))) " *
+                                    "and number of columns in " *
+                                    "matrix ($(size(mx, 2))) do not match"))
+        end
+        for (j, col) in enumerate(idxs)
+            sdf[!, col] = view(mx, :, j)
+        end
+        return sdf
```

Comment:
They are minimally different. The difference is that for `SubDataFrame` we will make a copy anyway when doing an assignment, so on RHS I do not make a copy.

Reply:
I see. Why not merge them and just use a branch for each type?

Reply:
I propose to leave it as-is for now. Otherwise we should move these definitions to abstractdataframe/abstractdataframe.jl where we avoided defining `setindex!` currently.

In the future, if we decide to refactor codebase so that we first define types in a single file e.g. `types.jl`, and then define methods for all types arranged cleanly (e.g. a separate file for `setindex!`) I would merge them.

I left a comment in the file that these methods should be merged in the future.